### PR TITLE
Adding new RKE test case: MultipleProjectMultipleNamespacesBackupAndRestore

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:
Added new tc:MultipleProjectMultipleNamespacesBackupAndRestore
**Which issue(s) this PR fixes** (optional)
Closes #PA-1498

**Special notes for your reviewer**:
log: https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/2241/consoleFull

P.S:  Will remove duplicate code from PR https://github.com/portworx/torpedo/pull/1602 once it is merged

### CODE REVIEW

File: tests/backup/backup_helper.go [hunk 1]
```diff
@@ -3910,6 +3911,50 @@ func CreateRestoreWithProjectMapping(restoreName string, backupName string, name
...
+	curBackups, err := backupDriver.EnumerateBackup(ctx, bkpEnumerateReq)
+	if err != nil {
+		return err
+	}
+	for _, bkp = range curBackups.GetBackups() {
+		if bkp.Name == backupName {
+			bkpUid = bkp.Uid
+			break
+		}
+	}
```
Comment (kphalgun-px) [inline tests/backup/backup_helper.go:3933]: How about using Inst().Backup.GetBackupUID?

File: tests/backup/backup_rancher_test.go [hunk 2]
```diff
@@ -265,3 +271,534 @@ var _ = Describe("{SingleNamespaceBackupRestoreToNamespaceInSameAndDifferentProj
...
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			restoreName = fmt.Sprintf("%s-replace-ns-project-move-%v", restoreNamePrefix, RandomString(5))
+			restoreList = append(restoreList, restoreName)
+			err = CreateRestoreOnRancherWithoutCheck(restoreName, backupName, namespaceMapping, destinationClusterName, orgID, ctx, nil, projectUIDMapping, projectNameMapping, 2)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Creating restore- %s with replace policy from backup %s", restoreName, backupName))
+			log.Infof("Move namespaces in destination cluster to diff project while restore is going on")
+			err = Inst().S.(*rke.Rancher).ChangeProjectForNamespace(destProjectList[1], restoreNamespaceList)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Moving namespaces %v from project %s to project %s", restoreNamespaceList, destProjectList[0], destProjectList[1]))
+			log.Infof("Verifying if restore is successful after moving destination namespaces from one project to another")
+			err = restoreSuccessCheck(restoreName, orgID, maxWaitPeriodForRestoreCompletionInMinute*time.Minute, restoreJobProgressRetryTime*time.Minute, ctx)
```
Comment (kphalgun-px) [inline tests/backup/backup_rancher_test.go:422]: Is it not restoreJobProgressRetryTime*time.Second and not restoreJobProgressRetryTime*time.Minute?

File: tests/backup/backup_rancher_test.go [hunk 3]
```diff
@@ -265,3 +271,534 @@ var _ = Describe("{SingleNamespaceBackupRestoreToNamespaceInSameAndDifferentProj
...
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Getting Project ID for project %s", sourceProject))
+				sourceProjectList = append(sourceProjectList, sourceProject)
+				sourceProjectIDList = append(sourceProjectIDList, sourceProjectID)
+			}
+		})
+
+		Step("Adding namespaces to project in source cluster and taking backup", func() {
+			log.InfoD("Adding namespaces to project in source cluster and taking backup")
+			log.Infof("Adding half of the namespaces to first project in source cluster")
+			err = Inst().S.(*rke.Rancher).AddNamespacesToProject(sourceProjectList[0], appNamespaces[0:len(appNamespaces)/2])
```
Comment (kphalgun-px) [inline tests/backup/backup_rancher_test.go:590]: Can we define a new variable for appNamespaces[0:len(appNamespaces)/2] and use it everywhere else?

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: drivers/scheduler/rke/rke.go [hunk 4]
```diff
@@ -270,6 +272,53 @@ func (r *Rancher) SaveSchedulerLogsToFile(n node.Node, location string) error {
...
 }
 
+// RemoveNamespaceFromProject moves namespace to no project
+func (r *Rancher) RemoveNamespaceFromProject(nsList []string) error {
+	for _, ns := range nsList {
+		ns, err := core.Instance().GetNamespace(ns)
+		if err != nil {
+			return err
+		}
+		delete(ns.Labels, "field.cattle.io/projectId")
```
Comment (kshithijiyer-px) [inline drivers/scheduler/rke/rke.go:282]: Would it be a good idea to use a loop here?  where you just fetch ns.labels and ns.annotatios and keep just one delete call inside it?

File: drivers/scheduler/rke/rke.go [hunk 5]
```diff
@@ -270,6 +272,53 @@ func (r *Rancher) SaveSchedulerLogsToFile(n node.Node, location string) error {
...
+	for _, ns := range nsList {
+		ns, err := core.Instance().GetNamespace(ns)
+		if err != nil {
+			return err
+		}
+		projectId, err = r.GetProjectID(projectName)
+		if err != nil {
+			return err
+		}
+		delete(ns.Labels, "field.cattle.io/projectId")
```
Comment (kshithijiyer-px) [inline drivers/scheduler/rke/rke.go:306]: Same suggestion

File: tests/backup/backup_helper.go (hunk 1 above)
Comment (snigdha-px) [reply tests/backup/backup_helper.go:3933]: What value will it add?
@kphalgun-px

File: tests/backup/backup_rancher_test.go (hunk 2 above)
Comment (snigdha-px) [reply tests/backup/backup_rancher_test.go:422]: Value of restoreJobProgressRetryTime is 1.
Checking after every 1 sec will not make sense.
@kphalgun-px ?

File: tests/backup/backup_rancher_test.go (hunk 3 above)
Comment (snigdha-px) [reply tests/backup/backup_rancher_test.go:590]: done

File: drivers/scheduler/rke/rke.go (hunk 4 above)
Comment (snigdha-px) [reply drivers/scheduler/rke/rke.go:282]: done

File: drivers/scheduler/rke/rke.go (hunk 5 above)
Comment (snigdha-px) [reply drivers/scheduler/rke/rke.go:306]: done

File: tests/backup/backup_rancher_test.go (hunk 2 above)
Comment (kphalgun-px) [reply tests/backup/backup_rancher_test.go:422]: My bad. I assumed the retry intervals are usually set in seconds.

File: tests/backup/backup_helper.go (hunk 1 above)
Comment (kphalgun-px) [reply tests/backup/backup_helper.go:3933]: Nothing. Only for the sake of brevity.

File: tests/backup/backup_rancher_test.go [hunk 6]
```diff
@@ -474,3 +479,327 @@ var _ = Describe("{NamespaceMoveFromProjectToProjectToNoProjectWhileRestore}", f
...
+				sourceProjectList = append(sourceProjectList, sourceProject)
+				sourceProjectIDList = append(sourceProjectIDList, sourceProjectID)
+			}
+		})
+
+		Step("Adding namespaces to project in source cluster and taking backup", func() {
+			log.InfoD("Adding namespaces to project in source cluster and taking backup")
+			log.Infof("Adding half of the namespaces to first project in source cluster")
+			halfProjectList := appNamespaces[0 : len(appNamespaces)/2]
+			err = Inst().S.(*rke.Rancher).AddNamespacesToProject(sourceProjectList[0], halfProjectList)
```
Comment (kshithijiyer-px) [inline tests/backup/backup_rancher_test.go:591]: Should we put this in a loop?

File: tests/backup/backup_rancher_test.go [hunk 7]
```diff
@@ -474,3 +479,327 @@ var _ = Describe("{NamespaceMoveFromProjectToProjectToNoProjectWhileRestore}", f
...
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Adding namespaces %s to project %s in source cluster", appNamespaces[len(appNamespaces)/2:], sourceProjectList[1]))
+			err = Inst().S.(*rke.Rancher).ValidateProjectOfNamespaces(sourceProjectList[1], appNamespaces[len(appNamespaces)/2:])
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if namespaces %s are added to project %s", appNamespaces[len(appNamespaces)/2:], sourceProjectList[1]))
+			projectNamespaces[sourceProjectList[1]] = appNamespaces[len(appNamespaces)/2:]
+			log.Infof("The value of project to namespace mapping in source cluster is %v", projectNamespaces)
+			fewNamespaceFromEachProject = append(fewNamespaceFromEachProject, appNamespaces[len(appNamespaces)/2:len(appNamespaces)/2+len(appNamespaces)/4]...)
+			log.Infof("The list of few namespaces from both the projects are %v:", fewNamespaceFromEachProject)
+			log.InfoD("Taking backup of all applications %s from both the projects in source cluster", appNamespaces)
+			backupName = fmt.Sprintf("%s-%v-all-ns", BackupNamePrefix, RandomString(10))
+			appContextsToBackup := FilterAppContextsByNamespace(scheduledAppContexts, appNamespaces)
```
Comment (kshithijiyer-px) [inline tests/backup/backup_rancher_test.go:608]: Should we put this in a loop?

**Review by kshithijiyer-px — CHANGES_REQUESTED** [review]
I see same blocks of code getting repeated twice can we move this into loops to reduce redundancy and make the code clean

File: tests/backup/backup_rancher_test.go (hunk 6 above)
Comment (snigdha-px) [reply tests/backup/backup_rancher_test.go:591]: done

**Comment by snigdha-px** [comment]
> I see same blocks of code getting repeated twice can we move this into loops to reduce redundancy and make the code clean

done

File: drivers/scheduler/rke/rke.go [hunk 8]
```diff
@@ -67,6 +68,7 @@ func (r *Rancher) GetRancherClusterParametersValue() (*RancherClusterParameters,
 	var rkeToken string
 	// TODO Rancher URL for cloud cluster will not be fetched from master node IP
 	masterNodeName := node.GetMasterNodes()[0].Name
+	log.Infof(" The master node here is %v", masterNodeName)
```
Comment (mkoppal-px) [inline drivers/scheduler/rke/rke.go:71]: nit: remove the space after double quotes

File: tests/backup/backup_rancher_test.go [hunk 9]
```diff
@@ -474,3 +479,323 @@ var _ = Describe("{NamespaceMoveFromProjectToProjectToNoProjectWhileRestore}", f
 		CleanupCloudSettingsAndClusters(backupLocationMap, credName, credUid, ctx)
 	})
 })
+
+// This testcase takes backup and restore of multiple namespaces belonging to multiple projects
+var _ = Describe("{MultipleProjectMultipleNamespacesBackupAndRestore}", func() {
```
Comment (mkoppal-px) [inline tests/backup/backup_rancher_test.go:484]: Can we change the test case name to `MultipleProjectsAndNamespacesBackupAndRestore` ?

**Review by mkoppal-px — CHANGES_REQUESTED** [review]
Can we use restore with validation?

File: drivers/scheduler/rke/rke.go (hunk 8 above)
Comment (snigdha-px) [reply drivers/scheduler/rke/rke.go:71]: done

File: tests/backup/backup_rancher_test.go (hunk 9 above)
Comment (snigdha-px) [reply tests/backup/backup_rancher_test.go:484]: done

**Comment by snigdha-px** [comment]
> Can we use restore with validation?

done

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Comment by snigdha-px** [comment]
> LGTM

https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/2273/consoleFull

**Pull request merged**